core/lib/mini_merkle_tree/src/lib.rs

```diff
@@ -86,67 +118,135 @@ where
+        }
+    }
+
+    /// Adds a new leaf to the tree (replaces leftmost empty leaf).
+    /// If the tree is full, its size is doubled.
+    /// Note: empty leaves != zero leaves.
+    pub fn push(&mut self, leaf: [u8; LEAF_SIZE]) {
+        let leaf_hash = self.hasher.hash_bytes(&leaf);
+        self.push_hash(leaf_hash);
+    }
+
+    /// Caches the rightmost `count` leaves.
+    /// Does not affect the root hash, but makes it impossible to get the paths to the cached leaves.
+    /// # Panics
+    /// Panics if `count` is greater than the number of non-cached leaves in the tree.
+    pub fn trim_start(&mut self, count: usize) {
+        assert!(self.hashes.len() >= count, "not enough leaves to cache");
+        let mut new_cache = vec![];
+        // Cache is a subset of the path to the first untrimmed leaf.
+        let root = self.compute_merkle_root_and_path(count, None, Some(&mut new_cache));
+        self.hashes.drain(..count);
+        self.start_index += count;
+        // It is important to add the root in case we just trimmed all leaves *and*
+        // the tree will grow on the next push.
+        new_cache.push(root);
+        self.cache = new_cache;
     }
 
     fn compute_merkle_root_and_path(
-        self,
-        mut index: usize,
-        mut merkle_path: Option<&mut Vec<H256>>,
+        &self,
+        mut end_index: usize,
+        mut start_path: Option<&mut Vec<H256>>,
+        mut end_path: Option<&mut Vec<H256>>,
     ) -> H256 {
-        assert!(index < self.hashes.len(), "invalid tree leaf index");
-
         let depth = tree_depth_by_size(self.binary_tree_size);
-        if let Some(merkle_path) = merkle_path.as_deref_mut() {
-            merkle_path.reserve(depth);
+        if let Some(left_path) = start_path.as_deref_mut() {
+            left_path.reserve(depth);
+        }
+        if let Some(right_path) = end_path.as_deref_mut() {
+            right_path.reserve(depth);
         }
 
-        let mut hashes = self.hashes;
-        let mut level_len = hashes.len();
+        let mut hashes = self.hashes.clone();
+        let mut start_index = self.start_index;
+
         for level in 0..depth {
             let empty_hash_at_level = self.hasher.empty_subtree_hash(level);
-
-            if let Some(merkle_path) = merkle_path.as_deref_mut() {
-                let adjacent_idx = index ^ 1;
-                let adjacent_hash = if adjacent_idx < level_len {
-                    hashes[adjacent_idx]
-                } else {
-                    empty_hash_at_level
-                };
-                merkle_path.push(adjacent_hash);
+            if start_index % 2 == 1 {
+                hashes.push_front(self.cache[level]);
+            }
+            if hashes.len() % 2 == 1 {
+                hashes.push_back(empty_hash_at_level);
             }
 
-            for i in 0..(level_len / 2) {
+            let push_sibling_hash = |path: Option<&mut Vec<H256>>, index: usize| {
```

Comment:
Bikeshedding (sort of): IMO, this is still quite complicated. I've come up with this alternative (doesn't include changes discussed in other line comments):

```rust
let mut hashes = self.hashes.clone();
let mut absolute_start_index = self.start_index;

for level in 0..depth {
    let empty_hash_at_level = self.hasher.empty_subtree_hash(level);
    let mut start_index = 0; // relative index in `hashes`
    if absolute_start_index % 2 == 1 {
        hashes.push_front(self.cache[level]);
        start_index += 1;
        end_index += 1;
    }
    if hashes.len() % 2 == 1 {
        hashes.push_back(empty_hash_at_level);
    }

    // Probably redundant (see other comments)
    if let Some(path) = start_path.as_deref_mut() {
        path.push(hashes[start_index ^ 1]);
    }
    if let Some(path) = end_path.as_deref_mut() {
        let hash = hashes.get(end_index ^ 1).copied().unwrap_or_default(); // looks bad; see the other comment
        path.push(hash);
    }

    let level_len = hashes.len() / 2;
    for i in 0..level_len {
        hashes[i] = self.hasher.compress(&hashes[2 * i], &hashes[2 * i + 1]);
    }

    hashes.truncate(level_len);
    end_index /= 2;
    absolute_start_index /= 2;
}
```

The idea is to use relative indexing everywhere, other than when pushing `cache`d entries.